arbnode/forwarder.go

```diff
@@ -186,3 +205,162 @@ func (f *TxDropper) StopAndWait() {}
+		f.fallbackTarget != "" && f.fallbackTarget != f.currentTarget
+}
+
+func (f *RedisTxForwarder) noError() time.Duration {
+	f.errors = 0
+	return f.config.UpdateInterval
+}
+
+func (f *RedisTxForwarder) getForwarder() *TxForwarder {
+	f.mtx.RLock()
+	defer f.mtx.RUnlock()
+	return f.forwarder
+}
+
+func (f *RedisTxForwarder) setForwarder(forwarder *TxForwarder) {
+	f.mtx.Lock()
+	defer f.mtx.Unlock()
+	f.forwarder = forwarder
+}
+
+// not thread safe vs initialize and itself
+func (f *RedisTxForwarder) update(ctx context.Context) time.Duration {
+	nextUpdateIn := f.noError
+	var newSequencerUrl string
+	var redisErr error
+	if f.redisCoordinator != nil {
+		newSequencerUrl, redisErr = f.redisCoordinator.CurrentChosenSequencer(ctx)
+		if redisErr != nil || newSequencerUrl == "" {
+			if f.shouldFallbackToStatic() && f.fallbackTarget != "" {
+				log.Warn("coordinator failed to find live sequencer, falling back to static url", "err", redisErr, "fallback", f.fallbackTarget)
+				newSequencerUrl = f.fallbackTarget
+				nextUpdateIn = f.retryAfterError
+			} else {
+				log.Warn("coordinator failed to find live sequencer", "err", redisErr)
+				return f.retryAfterError()
+			}
+		}
+	} else {
+		if f.fallbackTarget != "" {
+			log.Warn("redis coordinator not initialized, falling back to static url", "fallback", f.fallbackTarget)
+			newSequencerUrl = f.fallbackTarget
+		} else {
+			// TODO panic? - there is no way to recover from this point
+			log.Error("redis coordinator not initilized, no fallback available")
+			return f.retryAfterError()
+		}
+	}
+	if newSequencerUrl == f.currentTarget {
+		return nextUpdateIn()
+	}
+	var newForwarder *TxForwarder
+	for {
+		newForwarder = NewForwarder(newSequencerUrl, f.config)
+		err := newForwarder.Initialize(ctx)
+		if err == nil {
+			break
+		}
+		if f.shouldFallbackToStatic() && newSequencerUrl != f.fallbackTarget {
+			log.Error("failed to initialize forward agent, falling back to static url", "err", err, "fallback", f.fallbackTarget)
+			newSequencerUrl = f.fallbackTarget
+			nextUpdateIn = f.retryAfterError
+		} else {
+			log.Error("failed to initialize forward agent", "err", err)
+			return f.retryAfterError()
+		}
+	}
+	oldForwarder := f.getForwarder()
+	if oldForwarder != nil {
+		oldForwarder.Disable()
+	}
+	f.currentTarget = newSequencerUrl
+	f.setForwarder(newForwarder)
+	return nextUpdateIn()
+}
+
+func (f *RedisTxForwarder) Start(ctx context.Context) error {
+	if err := f.StopWaiterSafe.Start(ctx, f); err != nil {
+		return err
+	}
+	if err := f.CallIteratively(f.update); err != nil {
+		return errors.Wrap(err, "failed to start forwarder update thread")
+	}
+	return nil
+}
+
+func (f *RedisTxForwarder) StopAndWait() {
+	err := f.StopWaiterSafe.StopAndWait()
+	if err != nil {
+		log.Error("Failed to stop forwarder", "err", err)
+	}
+	// TODO is disabling of old forwarder required?
+	oldForwarder := f.getForwarder()
+	if oldForwarder != nil {
+		oldForwarder.Disable()
```

Comment:
Instead of Disable we should call StopAndWait here. Probably not necessary but it'd better fit the API.